question2/odbcdemo.c

```diff
@@ -0,0 +1,288 @@
+    }
+
+    printf("Driver version: %s\n", ver);
+    printf("Calling SQLGetTypeInfo...\n");
+    rc = SQLGetTypeInfo((SQLHSTMT)hstmt, SQL_ALL_TYPES);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLGetTypeInfo has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc = SQLBindCol(hstmt, 1, SQL_C_CHAR, &dataStruct[0].charCol1[0],
+                    (SDWORD)sizeof(dataStruct[0].charCol1),
+                    (SQLLEN *)&dataStruct[0].length1);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLBindCol(1) has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc = SQLBindCol(hstmt, 4, SQL_C_CHAR, &dataStruct2[0].charCol1[0],
+                    (SDWORD)sizeof(dataStruct2[0].charCol1),
+                    (SQLLEN *)&dataStruct2[0].length1);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLBindCol(4) has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc = SQLBindCol(hstmt, 5, SQL_C_CHAR, &dataStruct3[0].charCol1[0],
+                    (SDWORD)sizeof(dataStruct3[0].charCol1),
+                    (SQLLEN *)&dataStruct3[0].length1);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLBindCol(5) has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    // Set to bogus
+    dataStruct[0].length1 = -1;
+    dataStruct2[0].length1 = -1;
+    dataStruct3[0].length1 = -1;
+
+    int count = 0;
+
+    printf("Fetching result set...\n");
+    while (SQLFetch(hstmt) == SQL_SUCCESS) {
+        char *ptr;
+
+        if (count == 0) {
+            printf("LITERAL PREFIX\n");
+            printf("indicator(d): %d\n", dataStruct2[0].length1);
+            printf("indicator(lld): %lld\n", dataStruct2[0].length1);
+            printf("indicator(hex): ");
+
+            /* assign to char so we can dump out each byte */
+            ptr = (char *)&dataStruct2[0].length1;
+
+            for (i = 0; i < 8; i++) {
+                printf("%x", ptr[i]);
+            }
+            printf("\n");
+            printf("%-32s\n\n", dataStruct2[0].length1 == SQL_NULL_DATA
+                                    ? (UCHAR *)"NULL"
+                                    : dataStruct2[0].charCol1);
+
+            printf("LITERAL SUFFIX\n");
+            printf("indicator(d): %d\n", dataStruct3[0].length1);
+            printf("indicator(lld): %lld\n", dataStruct3[0].length1);
+            printf("indicator(hex): ");
+
+            /* assign to char so we can dump out each byte */
+            ptr = (char *)&dataStruct3[0].length1;
+
+            for (i = 0; i < 8; i++) {
+                printf("%x", ptr[i]);
+            }
```

Comment:
the size of long is implementation specific, so it isn't safe to blindly looping through 8 times
```suggestion
            for (i = 0; i < sizeof(dataStruct3[0].length1); i++) {
                printf("%x", ptr[i]);
            }
```